Enforce AirBNB style guide rules

akrolsmir opened the pull request:
We mention in the contributing docs that we expect the [AirBNB style guide for Javascript](https://github.com/airbnb/javascript), but this wasn't actually enforced in our eslintrc.

This PR adds the base AirBNB style rules, with some exceptions for the ones that we're currently violating and might require some discussion.

Also note that this doesn't yet include the [AirBNB react style rules](https://www.npmjs.com/package/eslint-config-airbnb-typescript)

akrolsmir committed:
Require AirBNB style guide (without React)

akrolsmir committed:
Automatic style fixes (from eslint --fix)

akrolsmir committed:
Manual style fixes, and disable often-violated rules

akrolsmir committed:
Enable no-underscore-dangle rule

akrolsmir committed:
Enable class-methods-use-this (mark methods as static)

akrolsmir committed:
Enable no-nested-ternary

akrolsmir committed:
Enable radix rule

We already pass radix=10 in our codebase e.g. Radio.tsx

karriebear commented on frontend/.eslintrc:
```diff
@@ -70,7 +72,28 @@
+    "no-shadow": "off",
+    "no-param-reassign": "off",
+    "no-plusplus": "off",
+    "no-console": "off",
```
I think we should keep the no-console on. We have our log.ts that everything should theoretically be using. We could just disable the rule within that file

akrolsmir replied:
Good point! I've turned on `no-console` everywhere but in our custom log.ts, as suggested.

akrolsmir committed:
Merge branch 'develop' into prettier

akrolsmir committed:
Disable eslint on third-party code

akrolsmir committed:
Permit this-less class method since MetricsManager loads a global

akrolsmir committed:
Enable no-console

Migrate existing console calls to use our custom logger

karriebear commented on frontend/.eslintrc:
```diff
@@ -70,7 +72,25 @@
+    // TODO: For each one, either fix and reenable, or provide a justification.
+    "import/extensions": "off",
+    "import/prefer-default-export": "off",
+    "import/no-extraneous-dependencies": "off",
```
i think`import/no-extraneous-dependencies` should probably be left on. not sure why we would be importing something thats not in our package.json?

akrolsmir replied:
The name is a bit misleading; the two issues that were triggered here are that
1. Some test-only code is imported as a devDependency which ESLint was suspicious of
2. Some of our third-party imports don't have Typescript types, which ESLint seems to choke on

I've gone ahead and just eslint-ignored those, and reenabled this rule.

akrolsmir committed:
Fix duplicate import

akrolsmir committed:
Check import rules in tsx files

- Enable "import/extensions"
- Break import cycle in Components
- Fix App import name

akrolsmir committed:
Enable "import/no-extraneous-dependencies"

Mostly by ignoring current violations :P

akrolsmir committed:
Disable typescript checks for MetricsManager tests

karriebear commented on frontend/.eslintrc:
```diff
@@ -70,13 +72,48 @@
+    ],
+    // Disable a bunch of AirBNB rules we're currently in violation of.
+    // TODO: For each one, either fix and reenable, or provide a justification.
+    "import/prefer-default-export": "off",
```
I always hated this rule ☝️

akrolsmir replied:
Haha glad we agree!
